In [2]:
import pandas as pd
dflect = pd.read_csv('lectures.csv', sep = ',')
dflect


lecture_id  tag  part           type_of
0            89  159     5           concept
1           100   70     1           concept
2           185   45     6           concept
3           192   79     5  solving question
4           317  156     5  solving question
..          ...  ...   ...               ...
413       32535    8     5  solving question
414       32570  113     3  solving question
415       32604   24     6           concept
416       32625  142     2           concept
417       32736   82     3           concept

[418 rows x 4 columns]

In [3]:
dflect.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   lecture_id  418 non-null    int64 
 1   tag         418 non-null    int64 
 2   part        418 non-null    int64 
 3   type_of     418 non-null    object
dtypes: int64(3), object(1)
memory usage: 13.2+ KB


Видим, что в датасете нет пропусков, все ячейки заполнены 

In [4]:
dflect.describe()

lecture_id         tag        part
count    418.000000  418.000000  418.000000
mean   16983.401914   94.480861    4.267943
std     9426.164660   53.586487    1.872424
min       89.000000    0.000000    1.000000
25%     9026.250000   50.250000    2.000000
50%    17161.500000   94.500000    5.000000
75%    24906.250000  140.000000    6.000000
max    32736.000000  187.000000    7.000000

столбец part принимает значение от 1 до 7

In [5]:
dflect.describe(include=['object'])

type_of
count       418
unique        4
top     concept
freq        222

всего 4 уникальных типа лекций, самый популярный - concept, он встречается 222 раза

In [6]:
dflect['type_of'].value_counts(normalize=True)

type_of
concept             0.531100
solving question    0.444976
intention           0.016746
starter             0.007177
Name: proportion, dtype: float64

starter - самый редкий тип, занимает менее 1% от общего числа, intention - занимает чуть более 1%, типы concept и solving question занимают почти равные доли - 44% и 53%, они встречаются чаще всего  

In [7]:
dfques = pd.read_csv('questions.csv', sep = ',')
dfques


question_id  bundle_id  correct_answer  part            tags
0                0          0               0     1   51 131 162 38
1                1          1               1     1       131 36 81
2                2          2               0     1  131 101 162 92
3                3          3               0     1  131 149 162 29
4                4          4               3     1    131 5 162 38
...            ...        ...             ...   ...             ...
13518        13518      13518               3     5              14
13519        13519      13519               3     5               8
13520        13520      13520               2     5              73
13521        13521      13521               0     5             125
13522        13522      13522               3     5              55

[13523 rows x 5 columns]

In [8]:
dfques.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   question_id     13523 non-null  int64 
 1   bundle_id       13523 non-null  int64 
 2   correct_answer  13523 non-null  int64 
 3   part            13523 non-null  int64 
 4   tags            13522 non-null  object
dtypes: int64(4), object(1)
memory usage: 528.4+ KB


Видим, что в датасете почти нет пропусков, исключение составляет колонка tags, в которой есть 1 пропуск

In [9]:
dfques.describe()

question_id     bundle_id  correct_answer          part
count  13523.00000  13523.000000    13523.000000  13523.000000
mean    6761.00000   6760.510907        1.455298      4.264956
std     3903.89818   3903.857783        1.149707      1.652553
min        0.00000      0.000000        0.000000      1.000000
25%     3380.50000   3379.500000        0.000000      3.000000
50%     6761.00000   6761.000000        1.000000      5.000000
75%    10141.50000  10140.000000        3.000000      5.000000
max    13522.00000  13522.000000        3.000000      7.000000

колонка part принимает значение от 1 до 7, значит в тесте было 7 разделов, на каждый вопрос было до 4 варинатов ответа

In [10]:
dfques.describe(include=['object'])

tags
count   13522
unique   1519
top         8
freq      738

тк в колонке tag есть значения записанные через пробел, мы не можем быть полностью уверены, по какой из лекций был самый частый вопрос, нужен более глубокий анализ

In [11]:
dftrain = pd.read_csv('train.csv', sep = ',')
dftrain

row_id  timestamp     user_id  content_id  content_type_id  \
0                  0          0         115        5692                0   
1                  1      56943         115        5716                0   
2                  2     118363         115         128                0   
3                  3     131167         115        7860                0   
4                  4     137965         115        7922                0   
...              ...        ...         ...         ...              ...   
101230327  101230327  428564420  2147482888        3586                0   
101230328  101230328  428585000  2147482888        6341                0   
101230329  101230329  428613475  2147482888        4212                0   
101230330  101230330  428649406  2147482888        6343                0   
101230331  101230331  428692118  2147482888        7995                0   

           task_container_id  user_answer  answered_correctly  \
0                          1            3                   1   
1                          2            2                   1   
2                          0            0                   1   
3                          3            0                   1   
4                          4            1                   1   
...                      ...          ...                 ...   
101230327                 22            0                   1   
101230328                 23            3                   1   
101230329                 24            3                   1   
101230330                 25            1                   0   
101230331                 26            3                   1   

           prior_question_elapsed_time prior_question_had_explanation  
0                                  NaN                            NaN  
1                              37000.0                          False  
2                              55000.0                          False  
3                              19000.0                          False  
4                              11000.0                          False  
...                                ...                            ...  
101230327                      18000.0                           True  
101230328                      14000.0                           True  
101230329                      14000.0                           True  
101230330                      22000.0                           True  
101230331                      29000.0                           True  

[101230332 rows x 10 columns]

In [12]:
dftrain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 101230332 entries, 0 to 101230331
Data columns (total 10 columns):
 #   Column                          Dtype  
---  ------                          -----  
 0   row_id                          int64  
 1   timestamp                       int64  
 2   user_id                         int64  
 3   content_id                      int64  
 4   content_type_id                 int64  
 5   task_container_id               int64  
 6   user_answer                     int64  
 7   answered_correctly              int64  
 8   prior_question_elapsed_time     float64
 9   prior_question_had_explanation  object 
dtypes: float64(1), int64(8), object(1)
memory usage: 7.5+ GB


В датасете находятся 8 видов данных челочиленного типа, 1 дробный и 1 тип "строка" 

In [13]:
dftrain.describe()

row_id     timestamp       user_id    content_id  \
count  1.012303e+08  1.012303e+08  1.012303e+08  1.012303e+08   
mean   5.061517e+07  7.703644e+09  1.076732e+09  5.219605e+03   
std    2.922268e+07  1.159266e+10  6.197163e+08  3.866359e+03   
min    0.000000e+00  0.000000e+00  1.150000e+02  0.000000e+00   
25%    2.530758e+07  5.243436e+08  5.408116e+08  2.063000e+03   
50%    5.061517e+07  2.674234e+09  1.071781e+09  5.026000e+03   
75%    7.592275e+07  9.924551e+09  1.615742e+09  7.425000e+03   
max    1.012303e+08  8.742577e+10  2.147483e+09  3.273600e+04   

       content_type_id  task_container_id   user_answer  answered_correctly  \
count     1.012303e+08       1.012303e+08  1.012303e+08        1.012303e+08   
mean      1.935222e-02       9.040624e+02  1.376123e+00        6.251644e-01   
std       1.377596e-01       1.358302e+03  1.192896e+00        5.225307e-01   
min       0.000000e+00       0.000000e+00 -1.000000e+00       -1.000000e+00   
25%       0.000000e+00       1.040000e+02  0.000000e+00        0.000000e+00   
50%       0.000000e+00       3.820000e+02  1.000000e+00        1.000000e+00   
75%       0.000000e+00       1.094000e+03  3.000000e+00        1.000000e+00   
max       1.000000e+00       9.999000e+03  3.000000e+00        1.000000e+00   

       prior_question_elapsed_time  
count                 9.887879e+07  
mean                  2.542381e+04  
std                   1.994815e+04  
min                   0.000000e+00  
25%                   1.600000e+04  
50%                   2.100000e+04  
75%                   2.966600e+04  
max                   3.000000e+05

видно, что все ячейки заполнены и отсутствуют пропуски
максимальное время между первым и последним взаимодействием пользователя составляет осталяет более 2,75 летc, среднее время - 89 дней
миниамльное среднее время ответа на вопрос составляет 0 секунд, максимальное 5 минут, в среднем пользователь тратил около 25 секунд на ответ

In [14]:
dftrain['user_id'].nunique()

393656

количество уникальных пользователей составляет 393656

In [15]:
dftrain.describe(include=['object'])

prior_question_had_explanation
count                       100837826
unique                              2
top                              True
freq                         89685560

In [16]:
dftrain['prior_question_had_explanation'].value_counts(normalize=True)

prior_question_had_explanation
True     0.889404
False    0.110596
Name: proportion, dtype: float64

В 89% пользователь видел правильный ответ на свой предыдущий вопрос

In [17]:
dftrain['answered_correctly'].value_counts(normalize=True)

answered_correctly
 1    0.644517
 0    0.336131
-1    0.019352
Name: proportion, dtype: float64

64% ответов на вопрос были верны

In [18]:
dftrain['content_type_id'].value_counts(normalize=True)

content_type_id
0    0.980648
1    0.019352
Name: proportion, dtype: float64

в 98% событий пользователь отвечал на овпросы, и только в 2% смотрел лекции